dhis-2/dhis-services/dhis-service-analytics/src/main/java/org/hisp/dhis/analytics/OrgUnitField.java

```diff
@@ -0,0 +1,266 @@
+import static org.hisp.dhis.analytics.util.AnalyticsSqlUtils.OWNERSHIP_TBL_ALIAS;
+import static org.hisp.dhis.analytics.util.AnalyticsSqlUtils.quote;
+import static org.hisp.dhis.program.AnalyticsType.EVENT;
+
+import lombok.EqualsAndHashCode;
+import lombok.Getter;
+
+import org.hisp.dhis.program.AnalyticsType;
+
+/**
+ * The organisation unit field to use for an event (or enrollment) analytics
+ * query. It may be one of the following types:
+ * <ul>
+ * <li>{@link OrgUnitFieldType#DEFAULT} use the orgUnit in the analytics table.
+ * OrgUnit structure columns (and group sets if needed) are included in the
+ * analytics table.</li>
+ * <li>{@link OrgUnitFieldType#ATTRIBUTE} use an event attribute or data element
+ * of type orgUnit. OrgUnit structure tables are joined to the attribute or data
+ * element column.</li>
+ * <li>{@link OrgUnitFieldType#REGISTRATION} use the TEI registration orgUnit.
+ * OrgUnit structure tables are joined to the analytics table registrationou
+ * column.</li>
+ * <li>{@link OrgUnitFieldType#ENROLLMENT use the program instance enrollment
+ * orgUnit. OrgUnit structure columns are included in the analytics table for
+ * enrollment queries. For event queries, orgUnit structure tables are joined to
+ * the enrolmentou column.</li>
+ * <li>{@link OrgUnitFieldType#OWNER_AT_START} use the TEI owner at the start of
+ * reporting period. OrgUnit columns are coalesced from the ownership analytics
+ * table when present, and from the enrollment orgUnit when not (the ou column
+ * for enrollment queries, or the enrollmentou column for event queries.)</li>
+ * <li>{@link OrgUnitFieldType#OWNER_AT_END} use the TEI owner at the end of
+ * reporting period. OrgUnit columns are coalesced from the ownership analytics
+ * table when present, and from the enrollment orgUnit when not (the ou column
+ * for enrollment queries, or the enrollmentou column for event queries.)</li>
+ * </ul>
+ *
+ * @author Jim Grace
+ */
+@Getter
+@EqualsAndHashCode
+public class OrgUnitField
+{
+    public static final OrgUnitField DEFAULT_ORG_UNIT_FIELD = new OrgUnitField( null );
+
+    private final String field;
+
+    private final OrgUnitFieldType type;
+
+    public OrgUnitField( String field )
+    {
+        this.field = field;
+
+        if ( isEmpty( field ) )
+        {
+            this.type = OrgUnitFieldType.DEFAULT;
+        }
+        else
+        {
+            if ( "REGISTRATIO".equals( field ) )
+            {
+                this.type = REGISTRATION;
+            }
+            else if ( "ENROLLMENTT".equals( field ) )
+            {
+                this.type = OrgUnitFieldType.ENROLLMENT;
+            }
+            else if ( "OWNERATSTAR".equals( field ) )
+            {
+                this.type = OrgUnitFieldType.OWNER_AT_START;
+            }
+            else if ( "OWNERATENDD".equals( field ) )
+            {
+                this.type = OrgUnitFieldType.OWNER_AT_END;
+            }
```

Comment:
Maybe better to add an alias field to the type and do a lookup of exact name first, if that fails lookup by alias or something similar? Or why not name the enums so that there is no need for this alias?

Reply:
Good spot. This was actually code used for testing that I forgot to remove. The real names for these options are matched by the Enum names and can be parsed with `Enum.getIfPresent()`. Since the front end doesn't yet generate these real names, but is useful for displaying the output of the analytics queries, I had created dummy data elements in my database with 11-character UIDs REGISTRATIO, ENROLLMENTT, OWNERATSTAR, and OWNERATENDD that could be selected by the frontend. I've now removed this code. Thanks!